### ISSUE TITLE
Commit forced reasoning-budget transitions to Qwen decoder state

### ISSUE THREAD

**Issue by aosama** [aosama/astronomical#260]
## Goal

Make a hard reasoning-token budget alter the model's actual autoregressive history before final-answer generation. When the Qwen3.5 reasoning budget is exhausted, Astronomical must feed a model-appropriate transition and end-of-thinking sequence through the normal decoder path instead of changing only the token exposed to the output parser.

## Current behavior

At revision [`9fa148c`](https://github.com/aosama/astronomical/tree/9fa148c9770f0aeaf8522e06ebc8e149368f09d7), Qwen3.5 execution:

1. Synchronizes the model-selected token.
2. Uses that original token to compute and asynchronously submit its successor.
3. Calls `build_generated_token_emission` afterward.
4. Replaces the externally emitted boundary token with `think_end_token_id` when the budget is reached.

The ordering is visible in [`advance_generation.rs`](https://github.com/aosama/astronomical/blob/9fa148c9770f0aeaf8522e06ebc8e149368f09d7/crates/model-serving/src/qwen3_5/inference_execution/advance_generation.rs#L269-L295) and [`advance_generation.rs`](https://github.com/aosama/astronomical/blob/9fa148c9770f0aeaf8522e06ebc8e149368f09d7/crates/model-serving/src/qwen3_5/inference_execution/advance_generation.rs#L334-L395). The substitution is implemented in [`generated_token_emission.rs`](https://github.com/aosama/astronomical/blob/9fa148c9770f0aeaf8522e06ebc8e149368f09d7/crates/model-serving/src/qwen3_5/inference_execution/generated_token_emission.rs#L93-L119), whose comment records that the key/value cache may retain drift after the override.

Consequently, the client and output parser observe an end-of-thinking token that the model never consumed. The decoder state retains the original reasoning token, and the already-submitted successor was conditioned on that original token. This can classify continued reasoning as visible answer text, weaken the transition into the final answer, and place reasoning-budget control and tool parsing on different token histories.

## Required behavior

A hard budget is an inference-time generation constraint, not a soft prompt hint. Preserve these semantics:

- `thinking_budget = N` permits at most `N` committed model-selected reasoning tokens.
- A natural reasoning end before `N` proceeds unchanged.
- Before the model can select reasoning token `N + 1`, generation enters a forced-transition state.
- Every forced transition token is selected before ordinary sampling, processed by the normal model forward, and committed to decoder state.
- Visible answer generation begins only after the complete reasoning-end sequence has been consumed by the model and output parser.
- `max_output_tokens` remains the total generated-position ceiling. Forced transition positions count toward that ceiling and context admission, but not toward the caller's reasoning-token budget.
- Requests that cannot fit the requested reasoning allowance, the complete forced transition, and at least one final-answer position fail validation with a bounded actionable error.

### 1. Move enforcement before sampling and successor submission

Remove reasoning-budget token replacement from output emission. The budget owner must decide whether the next position is ordinary reasoning or a forced transition before sampling logits and before any successor forward is submitted.

For a forced position, either:

- Mask all logits except the required transition token before greedy or sampled selection, matching established logits-processor implementations; or
- Inject the required token into the generated sequence before the next model forward.

Both approaches are valid only if the exact forced token becomes the decoder input and key/value-cache history. Do not emit one token while committing another.

### 2. Force a complete model-owned transition sequence

Represent budget enforcement as request-local state such as `Thinking`, `ForcingTransition { next_token_index }`, and `VisibleAnswer`. Resolve and validate the complete forced token sequence when the model is loaded or the request is prepared.

The sequence must support multiple tokens and remain owned by the concrete model family. Do not hardwire a universal token identifier. For Qwen-family behavior, use the validated reasoning marker contract and a transition equivalent to the Qwen recommendation:

```text
Considering the limited time by the user, I have to give the solution based on the thinking directly now.
</think>
```

The correctness requirement is that the model consumes an appropriate transition and closing sequence. The exact text and marker encoding may vary by validated model family or artifact contract.

### 3. Keep output parsing and generation state aligned

- Mark forced transition text as reasoning until the reasoning-end marker completes.
- Change to visible-answer mode only when the same committed token sequence also closes the output parser's reasoning state.
- Treat parser-supported implicit reasoning exits, including a tool-call transition, as natural exits so a budget sequence cannot be injected into tool-call arguments.
- Preserve prior reasoning content and interleaved tool-call history without exposing control markers as assistant-visible text.

### 4. Preserve deterministic and speculative paths

- Apply the same forced-token state before greedy and stochastic sampling.
- Prevent Multi-Token Prediction verification windows from crossing the remaining reasoning allowance or any forced-transition position.
- Disable draft acceptance while the forced sequence is being committed unless verification proves the exact forced sequence and decoder state are preserved.
- Cancellation, output limits, end-of-sequence handling, and performance attribution must remain request-local and bounded throughout the transition.

### 5. Make family support explicit

Qwen3.5 must use the corrected path. Any other advertised model family that accepts a positive `thinking_budget` must either implement equivalent model-consumed enforcement or reject the unsupported positive budget. Carrying a positive value without enforcing it is not sufficient.

## Acceptance criteria

- [ ] A controlled Qwen3.5 generation proves the model-selected boundary candidate is not committed when the budget is exhausted; the first forced transition token is committed instead.
- [ ] The decoder input history, key/value-cache history, emitted token history, and output-parser history contain the same ordered tokens across the boundary.
- [ ] A budget of `N` preserves exactly up to `N` ordinary reasoning tokens and begins the forced transition at the next generated position, without the current effective `N - 1` allowance.
- [ ] The complete multi-token transition and reasoning-end sequence is processed through ordinary model forwards before visible answer generation.
- [ ] Natural reasoning completion before the budget produces no injected transition.
- [ ] A tool call that implicitly ends reasoning cannot receive forced reasoning-end tokens inside its function name or arguments.
- [ ] Greedy, stochastic, target-only, and eligible Multi-Token Prediction paths produce the same boundary semantics.
- [ ] Requests without enough output/context positions for the reasoning allowance, transition sequence, and one visible-answer position fail before generation.
- [ ] `thinking_budget = 0` continues to disable thinking, and an omitted budget continues to use the model's normal thinking behavior.
- [ ] Laguna and future reasoning families either enforce positive budgets through a model-consumed family-owned path or reject them explicitly.
- [ ] Enabled performance attribution records start, end, elapsed time, outcome, and forced-token count for the budget-transition operation without adding disabled-path clock or file work.
- [ ] Hermetic behavior tests and a Romeo and Juliet model-artifact journey cover the complete user-visible reasoning-to-answer transition with live progress and a timeout no greater than 120 seconds.

## References

- [Qwen3 technical report, Thinking Budget](https://arxiv.org/html/2505.09388): when the threshold is reached, Qwen manually halts reasoning, inserts a stop-thinking instruction plus `</think>`, and then generates the final answer from the accumulated reasoning.
- [Qwen3 official thinking-budget example](https://github.com/QwenLM/Qwen3/blob/main/docs/source/getting_started/thinking_budget.md): generates bounded reasoning first, appends the reasoning and transition to the continuation context, and then generates the answer.
- [vLLM thinking-budget control](https://docs.vllm.ai/en/stable/features/reasoning_outputs/#thinking-budget-control): counts reasoning tokens and forces the configured `reasoning_end_str`; the end string may include a transition phrase.
- [vLLM thinking-budget state](https://docs.vllm.ai/en/stable/api/vllm/v1/sample/thinking_budget_state/): applies the forced reasoning-end tokens to logits before sampling, including speculative-decoding handling.
- [NVIDIA NIM thinking-budget control](https://docs.nvidia.com/nim/large-language-models/latest/thinking-budget-control.html): uses a logits processor to force an early-exit string and distinguishes a hard token ceiling from advisory reasoning effort.
- [llama.cpp reasoning-budget sampler](https://github.com/ggml-org/llama.cpp/blob/master/common/reasoning-budget.cpp): masks every non-required logit while emitting the complete forced sequence and advances state as those tokens are accepted.
- [MLX-VLM thinking-budget implementation](https://github.com/Blaizzy/mlx-vlm/pull/789): forces a reasoning-closing sequence and feeds the selected forced token into subsequent generation.

## Non-goals

- Mapping qualitative `reasoning_effort` values to token counts.
- Hiding reasoning that the model still generates internally.
- Truncating the request when the thinking budget is reached.
- Using an output-only parser rewrite as a substitute for model-consumed generation control.

**Comment by aosama** [comment]
Implementation is complete in PR #261.

The change now:

- Commits every forced Qwen3.5 transition token through the normal model forward and decoder/KV-cache history before visible answer generation.
- Preserves exactly the allowed ordinary reasoning tokens, natural reasoning exits, and tool-call boundaries.
- Bounds output admission and Multi-Token Prediction at the hard-budget boundary.
- Rejects positive Laguna budgets until that family has equivalent model-consumed enforcement.
- Records transition construction and exact forced-token attribution.

Verification includes the mandatory 16-step pre-commit suite, 246 Qwen3.5 hermetic tests, 8 direct-MLX MTP boundary tests, model-backed decoder-history qualification, and a smallest-configured-model public Chat Completions REST journey using the Romeo and Juliet fixture.

PR: https://github.com/aosama/astronomical/pull/261
